### PR TITLE
Fix deadlock when connection is simultaneously aborted and ended (#684).

### PR DESCRIPTION
**Pull request by cesarbs**
#684 

cc @halter73 @mikeharder @lodejard @davidfowl @benaadams

### CODE REVIEW

File: src/Microsoft.AspNetCore.Server.Kestrel/Http/Connection.cs [hunk 1]
```diff
@@ -198,12 +191,10 @@ public virtual void OnSocketClosed()
 
         private void ApplyConnectionFilter()
```
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Http/Connection.cs]: There was a very obvious deadlock here too: if `_connectionState != ConnectionState.CreatingFrame` we would call `ConnectionControl.End`, which would try to lock `_stateLock`.

I'm not sure if the lock is required only for synchronizing `_connectionState` or if everything else should be inside the lock, so to be safe I moved the `lock` statement to the `if` body after atomically updating `_connectionState`.

File: src/Microsoft.AspNetCore.Server.Kestrel/Http/Connection.cs [hunk 2]
```diff
@@ -35,7 +35,7 @@ public class Connection : ConnectionContext, IConnectionControl
         private readonly SocketOutput _rawSocketOutput;
 
         private readonly object _stateLock = new object();
-        private ConnectionState _connectionState;
+        private int _connectionState; // Interlocked won't work with enum
```
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Http/Connection.cs]: Let it be clear that I loathe having had to do this.

Comment (benaadams) [inline src/Microsoft.AspNetCore.Server.Kestrel/Http/Connection.cs]: And int casts? Shudder

Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Http/Connection.cs]: That would be much better, thanks.

**Comment by halter73** [comment]
I think [my suggestion](https://github.com/aspnet/KestrelHttpServer/issues/684#issuecomment-194594910) to simply take the lock inside the callback is simpler and better.

**Comment by cesarbs** [comment]
@halter73 But what about updating `_connectionState`? Would we do it inside the callback?

**Comment by halter73** [comment]
Exactly. You need to update the state with the lock. Doing the state change with just Interlocked.CompareExchange creates a race in `ApplyConnectionFilter`.

**Comment by cesarbs** [comment]
Updated.

I wasn't sure if you meant to move everything inside the callback, since that introduces a behavior change: whereas before `Abort()` would immediately update `_connectionState`, now it delays that until the callback is executed in another thread. I'm not sure what the implications of that are and I was afraid this could introduce more issues.

File: src/Microsoft.AspNetCore.Server.Kestrel/Http/Connection.cs [hunk 3]
```diff
@@ -152,23 +152,21 @@ public Task StopAsync()
 
         public virtual void Abort()
         {
-            lock (_stateLock)
+            // Frame.Abort calls user code while this method is always
+            // called from a libuv thread.
+            System.Threading.ThreadPool.QueueUserWorkItem(state =>
             {
```
Comment (halter73) [inline src/Microsoft.AspNetCore.Server.Kestrel/Http/Connection.cs]: Nit: Move `var connection = (Connection)state;` to the top and do `lock (connection._stateLock)`, `connection._connectionState`, etc...

File: src/Microsoft.AspNetCore.Server.Kestrel/Http/Connection.cs [hunk 4]
```diff
@@ -152,23 +152,21 @@ public Task StopAsync()
...
-                if (_connectionState == ConnectionState.CreatingFrame)
-                {
-                    _connectionState = ConnectionState.ToDisconnect;
-                }
-                else
```
Comment (halter73) [inline src/Microsoft.AspNetCore.Server.Kestrel/Http/Connection.cs]: Where did the else go!?

Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Http/Connection.cs]: Oops, fixed.

File: src/Microsoft.AspNetCore.Server.Kestrel/Http/Connection.cs (hunk 3 above)
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Http/Connection.cs]: Done.

**Comment by halter73** [comment]
:shipit:

**Pull request merged**